### PR TITLE
feat(composer): support stability modifiers (#6050)

### LINKED ISSUE #6050: Composer: Version constraints such as ^1.0@beta are not supported
**What Renovate type are you using?**

I'm self-hosting on GitLab inside a CI pipeline

**Describe the bug**

Composer [supports version constraints](https://getcomposer.org/doc/articles/versions.md#stability-constraints) such as `^1.0^RC` or `^2.0@beta`. These values are not supported by Renovate.

**Did you see anything helpful in debug logs?**

I see these messages in debug log:
```
 DEBUG: Dependency drupal/daterange_compact has unsupported value 1.x-dev (repository=axl-ks/ks)
DEBUG: Dependency drupal/field_permissions has unsupported value ^1.0@RC (repository=axl-ks/ks)
```

**To Reproduce**

In a repo, add any composer.json setup with version constraints such as the above.

**Additional context**

I think this is the case of just missing support of these style of version constraints available to composer.

### PR DESCRIPTION
**Pull request by hussainweb**
This PR adds support for parsing composer style stability specifiers as [described here](https://getcomposer.org/doc/04-schema.md#package-links). This is just my attempt to fix the related issue (more details in #6050) and I hope it is sufficient to even be able to parse and compare versions when retrieving from repositories correctly.

Closes #6050

### CODE REVIEW

File: lib/versioning/composer/index.spec.ts [hunk 1]
```diff
@@ -12,6 +12,11 @@ describe('semver.equals(a, b)', () => {
   it('should pad really short version', () => {
     expect(semver.equals('v1.0.0', '1')).toBe(true);
   });
+  it('should translate stability modifier', () => {
+    expect(semver.equals('1.0@alpha3', '1.0.0-alpha.3')).toBe(true);
```
Comment (skippednote) [inline lib/versioning/composer/index.spec.ts:16]: @hussainweb It shouldn't include the patch
```suggestion
    expect(semver.equals('1.0@alpha3', '1.0-alpha.3')).toBe(true);
```

File: lib/versioning/composer/index.ts [hunk 2]
```diff
@@ -39,6 +48,8 @@ function composer2npm(input: string): string {
   output = output.replace(/(?:^|\s)~([1-9][0-9]*(?:\.[0-9]*)?)(?: |$)/g, '^$1');
   // ~0.4 to >=0.4 <1
   output = output.replace(/(?:^|\s)~(0\.[1-9][0-9]*)(?: |$)/g, '>=$1 <1');
+  // 1.0@beta2 to 1.0-beta.2
+  output = composerStability2npm(output);
```
Comment (skippednote) [inline lib/versioning/composer/index.ts:52]: @hussainweb Can this just be
```suggestion
  output = output.replace(/[@|^]/, '-');
```

Comment (hussainweb) [reply lib/versioning/composer/index.ts:52]: I based this on [semver docs](https://docs.npmjs.com/misc/semver#prerelease-identifiers). The prerelease identifiers are of format `beta.5`. We will need to convert `beta5` to `beta.5` as well.

Comment (hussainweb) [reply lib/versioning/composer/index.ts:52]: @skippednote, is `1.0.0-beta5` a valid npm specifier?

File: lib/versioning/composer/index.spec.ts (hunk 1 above)
Comment (hussainweb) [reply lib/versioning/composer/index.spec.ts:16]: So, there is a check before which pads version string with zeroes. Although, now I see that that replacement may not be happening in this case.

**Comment by rarkins** [comment]
@hussainweb thanks for starting this PR. Can you add more tests that demonstrate the deeper usefulness of this change? e.g. functions like `maxSatisfyingVersion()` and `getNewValue()`. We want to ensure we don't end up with supporting a new syntax as "valid" but not actually being able to *update* it.

**Comment by hussainweb** [comment]
@rarkins, thanks for reviewing. I added tests around those two functions and hit edge cases in testing. Particularly, I have hit a block with a difference between how npm's semver and composer's semver handle differences.

The first thing to note here is that npm semver doesn't handle prerelease identifier on a short version. So, `4.0-beta` is not valid but `4.0.0-beta` is valid. This means that when someone writes `~4.0@beta` in a composer.json file, there is no suitable translation to npm's semver at all.

To see the differences, consider this:

```bash
$ ./node_modules/.bin/semver -r ">4.0.0-beta" 0.4.0 0.5.0 4.0.0-beta.1 4.0.0 4.2.0-beta 4.2.0-beta.2 4.2.0 5.0.0
4.0.0-beta.1
4.0.0
4.2.0
5.0.0
```

Even though I am asking for all versions greater than 4.0.0 with a prerelease of type beta, it doesn't return `4.2.0-beta`. This behaviour is different in composer which you can [compare here](https://semver.mwl.be/#!?package=drupal%2Fcore&version=%5E8.6&minimum-stability=beta). My constraint is `^8.6@beta` and it picks even `8.9.0-beta` releases.

I am not sure what I can do here as this needs fixing at a deeper level. Still, I think I can live with this for now while documenting it as a known issue. It's better than not supporting these versions at all.

@rarkins, what do you think?

**Comment by rarkins** [comment]
> @rarkins, what do you think?

I'm happy if this PR can make it *better* than before if nothing else is *worse*. Do you think this will result in any situations where users would think it's buggy behaviour?

**Comment by hussainweb** [comment]
@rarkins, I could be considered biased. :)

But my rationale is that support for this version style didn't exist at all, so there is no question of breaking that at least. As for the regular version numbers, I scanned different tests and I see a good number of cases covered. I am fairly confident in this change.

**Comment by hussainweb** [comment]
@rarkins, I'm going to add a few more tests to check the caret behaviour. There are fewer tests for that.

**Comment by hussainweb** [comment]
I created another PR (#6093) to add the tests that failed recently. Since that's failing as well on master, those failures have nothing to do with this change. We can talk about those failures in that PR. Maybe I am just not understanding rangeStrategy replace correctly?

**Comment by rarkins** [comment]
So when I say "worse", I guess the easiest way to put that is: will this change result in new PRs being created that are *wrong*? If the only PRs it adds are correct, then I'd consider that positive.

**Comment by hussainweb** [comment]
@rarkins, I am afraid I don't understand. Based on what I understand, I can say that this change adds support for a very common use-case (which is _better_). At the same time, it does not break anything existing as far as I can see. I don't see how this change will result in new _wrong_ PRs. Can you give me an example of a wrong PR so that I may understand better?

**Comment by rarkins** [comment]
I'll try to clarify what I meant by bad:

- If this change will result in some *incorrect* PRs to users (e.g. updating them to an invalid version)
- If this change will result in additional warnings and/or errors in logs

**Comment by hussainweb** [comment]
@rarkins, I understand now.

> If this change will result in some incorrect PRs to users

Insofar as the tests are passing, I don't think so. I added some more tests as well and there is no change of behaviour.

> If this change will result in additional warnings and/or errors in logs

Not really. There might be new warnings if the version is not valid but those would be generating a different warning/error even now. There are still some more formats composer supports which this PR doesn't address (e.g. `1.x-dev`, `3.4.7 as 4.3`). They will continue to generate same errors as now.

**Comment by rarkins** [comment]
@hussainweb I've merged master into this, including your other composer fixes. Please confirm again if you think this is ready for final review and merge

**Comment by hussainweb** [comment]
@rarkins, yes, it is ready for your review now. Those tests now give me greater confidence. Thanks for merging master into this branch.

**Pull request merged**